hamocc/mo_carchm.F90

```diff
@@ -399,43 +407,56 @@ subroutine carchm(kpie,kpje,kpke,kbnd,pdlxp,pdlyp,pddpo,prho,pglat,omask,psicomo
                 atnh3  = atm(i,j,iatmnh3)
               endif
 
-              ! Ratio P/P_0, where P is the local SLP and P_0 is standard pressure (1 atm). This is
-              ! used in all surface flux calculations where atmospheric concentration is given as a
-              ! mixing ratio (i.e. partial presure = mixing ratio*SLP/P_0 [atm])
-              rpp0 = ppao(i,j)/101325.0
+              ! Sea level pressure in atm. This is used in all surface flux calculations where 
+              ! atmospheric concentration is given as a mixing ratio (i.e. partial presure = mixing ratio*SLP/P_0 [atm])
+              rpp0  = ppao(i,j)/101325.0
+
+              ! calculate correction for non-ideality of CO2 (fugacity coefficient, Weiss and Price 1980)
+              Bvir  = -1636.75 + 12.0408*tk -0.0327957*tk**2 + 0.0000316528*tk**3
+              delta = 57.7-0.118*tk
+              fc    = exp( rpp0*(Bvir + 2.0*delta)/(82.057*tk) )
+
+              ! calculate water vapor pressure [atm] as function of temperature and salinity (Weiss and Price 1980)
+              pH2O  = exp(24.4543 - 67.4509*(100.0/tk) - 4.8489*log(tk/100.0) - 0.000544*s)
+
+              ! Calculate the CO2 concentration in equilibrium with atmospheric x' (atco2=mole fraction of CO2 in dry air [ppm])
+              cu_sat = Kh0*atco2*1e-6*(rpp0-pH2O)*fc
+
+              fluxd = cu_sat*kwco2*dtbgc*rrho ! cu_sat and cu are in mol/kg. Multiply by rrho (g/cm^3) 
+              fluxu = cu    *kwco2*dtbgc*rrho ! to get fluxes in kmol/m^2
+
+              ! Set limit for CO2 outgassing to avoid negative DIC concentration, set minimum DIC concentration to 1e-5 kmol/m3
+              fluxu = min(fluxu,fluxd-(1e-5 - ocetra(i,j,k,isco212))*pddpo(i,j,1))
```

Comment:
Here and below: Not sure, but it could be worth considering  parameter that defines the minimum value (could be added in `mo_param_bgc.F90`)

Reply:
I agree, but I'll leave this for later...